velox/docs/develop/simd.rst

```diff
@@ -0,0 +1,179 @@
+``mask[i]`` is set, otherwise use the element in ``src[i]``.  In other words,
+the function returns ``dst`` where
+
+::
+
+   if mask[i]
+     dst[i] = load(base + indices[i])
+   else
+     dst[i] = src[i]
+
+Bit Masks
+~~~~~~~~~
+
+As already mentioned above, ``batch_bool`` is used to represent the result of a
+comparison, and the underlying data can be either a bit mask or a SIMD vector.
+To allow us manipulate this result, we provides some utilities to convert
+between ``batch_bool`` and bit mask (``toBitMask`` and ``fromBitMask``).  Once
+you convert it to bit mask, you can use the normal bit manipulating operations
+on it.  We also provide utilities like ``leadingMask`` and ``allSetBitMask`` to
+make it easier and faster to manipulate bits.
+
+Filter
+~~~~~~
+
+Another important function we have in ``SimdUtil.h`` is ``filter``.  It takes a
+SIMD vector ``data`` and a ``bitMask``, then for each ``i`` where ``bitMask[i]``
+is set, we move the corresponding ``data[i]`` to front and return the result.
+This behaves very similar to ``std::partition``.  In other words, the function
+returns ``dst`` where
+
+::
+
+   j = 0
+   for i in 0 to n
+     if bitMask[i]
+       dst[j++] = data[i]
+   for i in 0 to n
+     if not bitMask[i]
+       dst[j++] = data[i]
+
+BMI Utilities
+-------------
+
+In addition to SIMD abstraction and utilities, we also have some functions that
+depends on BMI2 intrinsics.  We define the portable version of them in
+``common/base/BitUtil.h``.  These functions include ``extractBits`` and
+``rotateLeft``.  They are relatively simple and standalone comparing to SIMD,
+and you can refer the documentation in the file for their usage.
+
+Use Cases
+---------
+
+Hash Table
+~~~~~~~~~~
+
+In ``BigintValuesUsingHashTable::testValues`` we use SIMD to check whether
+multiple values are in the hash table at same time.  In the hash table we use a
+special empty marker to indicate the value is missing.  The process is
+following:
+
+1. If all values are out of range, we can return all false.
+2. If empty marker has been inserted into the hash table, fall back to check the
+   values one by one.
+3. Hash all valid values using SIMD multiplication and modulo, and then get the
+   corresponding states in hash table using ``maskGather``.
+4. If the state is empty marker, the value is missing; if the state is equal to
+   value, the value is found.  Otherwise we have a hash collision and need to
+   look at next positions in hash table.  If no collision is happening, we can
+   return the result right now.
+5. For each value that has collision, we use SIMD to advance multiple positions
+   at one time, until we find either value match or empty mark.
+
+Filtering
+~~~~~~~~~
+
+A typical use case for filtering values using SIMD is in ``processFixedFilter``
+from ``dwio/dwrf/common/DecoderUtil.h``.  Basically what this function does is
+to get some indices from ``loadIndices``, filtering on the values, and storing
```

Comment:
"to get" -> "getting"

Reply:
I this the description of this function can be more high level and be easier to understand. How about:
```
This function evaluates the filter on a batch of values simultaneously using SIMD comparison instructions, and stores the passed row numbers for this batch to the result row number array (filterHits)
```